android/src/main/java/org/whispersystems/signalservice/api/util/RealtimeSleepTimer.java

```diff
@@ -0,0 +1,82 @@
+package org.whispersystems.signalservice.api.util;
+
+import android.app.AlarmManager;
+import android.app.PendingIntent;
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.content.IntentFilter;
+import android.os.Build;
+import android.os.SystemClock;
+import android.util.Log;
+
+import org.whispersystems.signalservice.api.util.SleepTimer;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A sleep timer that is based on elapsed realtime, so
+ * that it works properly, even in low-power sleep modes.
+ *
+ */
+public class RealtimeSleepTimer implements SleepTimer {
+  private static final String TAG = RealtimeSleepTimer.class.getSimpleName();
+
+  private final AlarmReceiver alarmReceiver;
+  private final Context context;
+
+  public RealtimeSleepTimer(Context context) {
+    this.context = context;
+    alarmReceiver = new RealtimeSleepTimer.AlarmReceiver();
```

Comment:
Is there a reason this needs to be an instance variable?

Reply:
I'm not sure how to answer that.  Would you prefer for the caller of the `sleep` method to have to pass it as as an argument, or is there something obvious that I'm missing?

Reply:
Sorry, I mistook your question as referring to the `context` variable.  Regarding `alarmReceiver`, as far as I can remember, the reason I made it into an instance variable, was to avoid reinstantiating the class on every sleep call.  Apart from this consideration, I don't think it _needs_ to be an instance variable.